In [163]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [164]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "/Users/divya/Desktop/Module19/Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [165]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [166]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

### Prepare the Data

In [167]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()
scaler.fit(df_market_data)
df_market_data_scaled = scaler.transform(df_market_data)
df_market_data_scaled

array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [168]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(df_market_data_scaled, index=df_market_data.index, columns=df_market_data.columns)


# Copy the crypto names from the original data
df_market_data_scaled['coin_id'] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled.set_index('coin_id', inplace=True)

# Display sample data
df_market_data_scaled.head(10)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [169]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [170]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
    kmeans = KMeans(n_clusters=k)
    kmeans.fit(df_market_data_scaled)
    inertia_values.append(kmeans.inertia_)



In [171]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data1 = {'k': k_values, 'inertia': inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df1 = pd.DataFrame(elbow_data1)



In [172]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow1 = elbow_df1.hvplot(kind='line',
                             x='k',
                             y='inertia',
                             width=500,
                             title='Original',
                             xaxis={'ticks': list(elbow_df1['k'].unique())})
elbow1

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4 is the best value for 'k' because it is the point at which the curve starts to flatten out.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [173]:
# Initialize the K-Means model using the best value for k
kmeans = KMeans(n_clusters=4, random_state = 1)


In [174]:
# Fit the K-Means model using the scaled data
kmeans.fit(df_market_data_scaled)


KMeans(n_clusters=4, random_state=1)

In [175]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cluster_labels = kmeans.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
print(cluster_labels)


[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [176]:
# Create a copy of the DataFrame
dfcrypto_predict = df_market_data_scaled.copy()



In [177]:
# Add a new column to the DataFrame with the predicted clusters
dfcrypto_predict['cluster'] = cluster_labels

# Display sample data
dfcrypto_predict.head(10)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  cluster  
coin_id                                               
bitcoin                           -0.251637        2  
ethereum                          -0.199352        2  
tether                            -0.282061        0  
ripple                            -0.295546        0  
bitcoin-cash                      -0.270317        2  
binancecoin                       -0.225533        2  
chainlink                         -0.018284        2  
cardano                           -0.155428        2  
litecoin                          -0.292351        2  
bitcoin-cash-sv                   -0.206029        0

In [178]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
cluster_original = dfcrypto_predict.hvplot.scatter(x='price_change_percentage_24h', y='price_change_percentage_7d', by='cluster', hover_cols=['coin_id'])
cluster_original

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [179]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)


In [180]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
df_market_data_pca = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 
df_market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [181]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
explained_variance = pca.explained_variance_ratio_
print(explained_variance)

[0.3719856  0.34700813 0.17603793]


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.3719856 + 0.34700813 + 0.17603793 = 0.89503166
The total explained variance of the 3 principal components are 89.50%

In [182]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(df_market_data_pca, columns=['PC1', 'PC2', 'PC3'])

# Copy the crypto names from the original data
df_market_data_pca['coin_id'] = df_market_data.index


# Set the coinid column as index
df_market_data_pca.set_index('coin_id', inplace=True)


# Display sample data
df_market_data_pca.head(10)


PC1       PC2       PC3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the PCA Data

In [183]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 11))

In [184]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for k in k_values:
    kmeans = KMeans(n_clusters=k)
    kmeans.fit(df_market_data_pca)
    inertia_values.append(kmeans.inertia_)



In [185]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'K Values': k_values, 'Inertia': inertia_values}

# Create a DataFrame with the data to plot the Elbow curve
elbow2_df = pd.DataFrame(elbow_data)



In [186]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow2 = elbow2_df.hvplot(kind='line',
                         x='K Values',
                         y='Inertia',
                         width=500,
                         title='Elbow Curve PCA',
                         xlabel='K Values',
                         ylabel='Inertia')
elbow2

:Curve   [K Values]   (Inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4 because it is when the data starts to flatten out.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, but it is way clearer in this chart because there are more discrete points.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [187]:
# Initialize the K-Means model using the best value for k
kmeans = KMeans(n_clusters=4, random_state = 1)


In [188]:
# Fit the K-Means model using the PCA data
kmeans.fit(df_market_data_pca)


KMeans(n_clusters=4, random_state=1)

In [189]:
# Predict the clusters to group the cryptocurrencies using the PCA data
cluster_labels = kmeans.predict(df_market_data_pca)

# Print the resulting array of cluster values.
print(cluster_labels)


[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [190]:
# Create a copy of the DataFrame with the PCA data
dfpca_predict = df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
dfpca_predict['cluster'] = cluster_labels

# Display sample data
dfpca_predict.head(10)


PC1       PC2       PC3  cluster
coin_id                                               
bitcoin         -0.600667  0.842760  0.461595        2
ethereum        -0.458261  0.458466  0.952877        2
tether          -0.433070 -0.168126 -0.641752        0
ripple          -0.471835 -0.222660 -0.479053        0
bitcoin-cash    -1.157800  2.041209  1.859715        2
binancecoin     -0.516534  1.388377  0.804071        2
chainlink       -0.450711  0.517699  2.846143        2
cardano         -0.345600  0.729439  1.478013        2
litecoin        -0.649468  0.432165  0.600303        2
bitcoin-cash-sv -0.759014 -0.201200 -0.217653        0

In [191]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
cluster_pca = dfpca_predict.hvplot.scatter(x='PC1', y='PC2', by='cluster', hover_cols=['coin_id'])
cluster_pca

:NdOverlay   [cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [192]:
# Composite plot to contrast the Elbow curves
elbow1 + elbow2



:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [K Values]   (Inertia)

In [193]:
# Composite plot to contrast the clusters
cluster_original + cluster_pca



:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster]
      :Scatter   [PC1]   (PC2,coin_id)

In [194]:
pd.crosstab(dfcrypto_predict['cluster'], dfpca_predict['cluster'])

cluster   0  1   2  3
cluster              
0        26  0   0  0
1         0  1   0  0
2         0  0  13  0
3         0  0   0  1

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using fewer features to cluster the data using K-Means is that it reduces the dimensionality of the data, making it easier to visualize and analyze. It also helps to remove noise and irrelevant features, which can improve the accuracy of the clustering results. In this case, using PCA to reduce the dimensionality of the data resulted in more distinct and well-separated clusters.